### PR TITLE
Update of VS Tools for Docker doc

### PR DESCRIPTION
**Pull request by scottaddie**
Fixes https://github.com/aspnet/Docs/issues/6813
Fixes https://github.com/aspnet/Docs/issues/6128
Fixes https://github.com/aspnet/Docs/issues/5631

**Internal Review Pages**
* [2.1](https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/docker/visual-studio-tools-for-docker?view=aspnetcore-2.1&branch=pr-en-us-7772)
* [2.0](https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/docker/visual-studio-tools-for-docker?view=aspnetcore-2.0&branch=pr-en-us-7772)

### CODE REVIEW

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md [hunk 1]
```diff
@@ -4,40 +4,42 @@ author: spboyer
...
+* [Visual Studio 2017](https://www.visualstudio.com/) with the **.NET Core cross-platform development** workload
 
 ## Installation and setup
 
 For Docker installation, review the information at [Docker for Windows: What to know before you install](https://docs.docker.com/docker-for-windows/install/#what-to-know-before-you-install) and install [Docker For Windows](https://docs.docker.com/docker-for-windows/install/).
 
 **[Shared Drives](https://docs.docker.com/docker-for-windows/#shared-drives)** in Docker for Windows must be configured to support volume mapping and debugging. Right-click the System Tray's Docker icon, select **Settings...**, and select **Shared Drives**. Select the drive where Docker stores files. Select **Apply**.
 
-![Shared Drives](./visual-studio-tools-for-docker/_static/settings-shared-drives-win.png)
+![Shared Drives](visual-studio-tools-for-docker/_static/settings-shared-drives-win.png)
```
Comment (guardrex) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: Although a reader will read the screen, I recommend a more specific alt. Perhaps ...

> Dialog to select the local drives for containers with the C drive selected

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md [hunk 2]
```diff
@@ -4,40 +4,42 @@ author: spboyer
...
 
-* [Visual Studio 2017](https://www.visualstudio.com/) with the **.NET Core cross-platform development** workload
 * [Docker for Windows](https://docs.docker.com/docker-for-windows/install/)
+* [Visual Studio 2017](https://www.visualstudio.com/) with the **.NET Core cross-platform development** workload
 
 ## Installation and setup
 
 For Docker installation, review the information at [Docker for Windows: What to know before you install](https://docs.docker.com/docker-for-windows/install/#what-to-know-before-you-install) and install [Docker For Windows](https://docs.docker.com/docker-for-windows/install/).
 
 **[Shared Drives](https://docs.docker.com/docker-for-windows/#shared-drives)** in Docker for Windows must be configured to support volume mapping and debugging. Right-click the System Tray's Docker icon, select **Settings...**, and select **Shared Drives**. Select the drive where Docker stores files. Select **Apply**.
```
Comment (guardrex) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: If you want to ... drop ellipses from \*\*Settings...\*\*

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md [hunk 3]
```diff
@@ -46,53 +48,119 @@ If the target framework is .NET Core, the **OS** drop-down allows for the select
...
+
+## Dockerfile overview
+
+A *Dockerfile*, the recipe for creating a final Docker image, is added to the project root. Refer to [Dockerfile reference](https://docs.docker.com/engine/reference/builder/) for an understanding of the commands within it. This particular *Dockerfile* uses a [multi-stage build](https://docs.docker.com/engine/userguide/eng-image/multistage-build/) with four distinct, named build stages:
+
+::: moniker range=">= aspnetcore-2.1"
+
+[!code-dockerfile[](visual-studio-tools-for-docker/samples/2.1/HelloDockerTools/Dockerfile.original?highlight=1,6,14,17)]
+
+The preceding *Dockerfile* is based on the [microsoft/dotnet](https://hub.docker.com/r/microsoft/dotnet/) image. This base image includes the ASP.NET Core runtime and NuGet packages. The packages have been pre-jitted to improve startup performance.
```
Comment (guardrex) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: * "are"
* JIT seems to be spelled out on first reference (e.g., https://docs.microsoft.com/dotnet/standard/managed-execution-process)

> The packages are just-in-time (JIT) compiled to improve startup performance.

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md [hunk 4]
```diff
@@ -46,53 +48,119 @@ If the target framework is .NET Core, the **OS** drop-down allows for the select
...
+Visual Studio 2017 versions 15.8 or later add an orchestration solution only when instructed. Right-click the project in **Solution Explorer** and select **Add** > **Container Orchestrator Support**. Two different choices are offered: [Docker Compose](#docker-compose) and [Service Fabric](#service-fabric).
 
-The *Dockerfile* is based on the [microsoft/aspnetcore](https://hub.docker.com/r/microsoft/aspnetcore) image. This base image includes the ASP.NET Core NuGet packages, which have been pre-jitted to improve startup performance.
+### Docker Compose
 
-The *docker-compose.yml* file contains the name of the image that's created when the project runs:
+The Visual Studio Tools for Docker add a *docker-compose* project to the solution with the following files:
 
-[!code-yaml[](visual-studio-tools-for-docker/samples/HelloDockerTools/docker-compose.yml?highlight=5)]
+* *docker-compose.dcproj* &mdash; The file representing the project. Includes a `<DockerTargetOS>` element specifying the OS to be used.
```
Comment (guardrex) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: I was using `&ndash;` on these. I thought the long em dashes were  ... well ... *too long*. :smile:

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md [hunk 5]
```diff
@@ -46,53 +48,119 @@ If the target framework is .NET Core, the **OS** drop-down allows for the select
...
+* *.dockerignore* &mdash; Lists the file and directory patterns to exclude when generating a build context.
+* *docker-compose.yml* &mdash; The base [Docker Compose](https://docs.docker.com/compose/overview/) file used to define the collection of images to be built and run with `docker-compose build` and `docker-compose run`, respectively.
```
Comment (guardrex) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: Can the "to be" be removed?

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md [hunk 6]
```diff
@@ -46,53 +48,119 @@ If the target framework is .NET Core, the **OS** drop-down allows for the select
...
+
+The *docker-compose.yml* file references the name of the image that's created when the project runs:
+
+[!code-yaml[](visual-studio-tools-for-docker/samples/2.0/docker-compose.yml?highlight=5)]
 
 In the preceding example, `image: hellodockertools` generates the image `hellodockertools:dev` when the app runs in **Debug** mode. The `hellodockertools:latest` image is generated when the app runs in **Release** mode.
 
-Prefix the image name with the [Docker Hub](https://hub.docker.com/) username (for example, `dockerhubusername/hellodockertools`) if the image will be pushed to the registry. Alternatively, change the image name to include the private registry URL (for example, `privateregistry.domain.com/hellodockertools`) depending on the configuration.
+Prefix the image name with the [Docker Hub](https://hub.docker.com/) username (for example, `dockerhubusername/hellodockertools`) if the image is to be pushed to the registry. Alternatively, change the image name to include the private registry URL (for example, `privateregistry.domain.com/hellodockertools`) depending on the configuration.
```
Comment (guardrex) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: ditto ... can "to be" be dropped?

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md [hunk 7]
```diff
@@ -46,53 +48,119 @@ If the target framework is .NET Core, the **OS** drop-down allows for the select
...
+### Service Fabric
+
+In addition to the base [Prerequisites](#prerequisites), the [Service Fabric](/azure/service-fabric/) orchestration solution demands the following prerequisites:
+
+* [Microsoft Azure Service Fabric SDK](https://www.microsoft.com/web/handlers/webpi.ashx?command=getinstallerredirect&appid=MicrosoftAzure-ServiceFabric-CoreSDK) version 2.6 or later
+* Visual Studio 2017's **Azure Development** workload
+
+Service Fabric doesn't support running Linux containers in the local development cluster on Windows. If the project is already using a Linux container, Visual Studio prompts to switch to Windows containers.
+
+The Visual Studio Tools for Docker do the following tasks:
```
Comment (guardrex) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: Maybe "perform" over the more informal "do"

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md [hunk 8]
```diff
@@ -46,53 +48,119 @@ If the target framework is .NET Core, the **OS** drop-down allows for the select
...
+
+* Adds a *&lt;project_name&gt;Application* **Service Fabric Application** project to the solution.
+* Adds a *Dockerfile* and a *.dockerignore* file to the ASP.NET Core project. If a *Dockerfile* already exists in the ASP.NET Core project, it's renamed to *Dockerfile.original*. A new *Dockerfile*, similar to the following, is created:
+
+    [!code-dockerfile[](visual-studio-tools-for-docker/samples/2.1/HelloDockerTools/Dockerfile)]
+
+* Adds an `<IsServiceFabricServiceProject>` element to the ASP.NET Core project's *.csproj* file:
```
Comment (guardrex) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: I think we're bolding these bad boys ... elements in either markup or project config files. I guessed a little bit from our markup guidelines anyway. Let me know if I should stop doing it.

> an \*\*\&lt;IsServiceFabricServiceProject\&gt;\*\* element

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md [hunk 9]
```diff
@@ -46,53 +48,119 @@ If the target framework is .NET Core, the **OS** drop-down allows for the select
...
+For more information, see [Deploy a .NET app in a Windows container to Azure Service Fabric](/azure/service-fabric/service-fabric-host-app-in-a-container).
 
 ## Debug
 
 Select **Docker** from the debug drop-down in the toolbar, and start debugging the app. The **Docker** view of the **Output** window shows the following actions taking place:
 
+::: moniker range=">= aspnetcore-2.1"
+
+* The *2.1-aspnetcore-runtime* tag of the *microsoft/dotnet* runtime image is acquired (if not already in the cache). The image installs the ASP.NET Core and .NET Core runtimes and associated libraries. It's optimized for running ASP.NET Core apps in production.
+* The *ASPNETCORE_ENVIRONMENT* environment variable is set to `Development` within the container.
```
Comment (guardrex) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: Code fence env vars ...

> \`ASPNETCORE_ENVIRONMENT\`

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md [hunk 10]
```diff
@@ -46,53 +48,119 @@ If the target framework is .NET Core, the **OS** drop-down allows for the select
...
+* Two dynamically assigned ports are exposed: one for HTTP and one for HTTPS. The port assigned to localhost can be queried with the `docker ps` command.
+* The app is copied to the container.
+* The default browser is launched with the debugger attached to the container using the dynamically assigned port.
+
+The resulting Docker image of the app is tagged as *dev*. The image is based on the *2.1-aspnetcore-runtime* tag of the *microsoft/dotnet* base image. Run the `docker images` command in the **Package Manager Console** (PMC) window. The images on the machine are displayed:
+
+```console
+REPOSITORY          TAG                      IMAGE ID            CREATED             SIZE
```
Comment (guardrex) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: Three spots (here and below) ...

Do these blow the 80-85 char width we discussed previously? I see a small opportunity to narrow the columns here and below ... although ... it isn't by much. I agree tho with use of a code block ... and if it runs wide ... oh, well ... it is what it is. lol

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md [hunk 11]
```diff
@@ -46,53 +48,119 @@ If the target framework is .NET Core, the **OS** drop-down allows for the select
...
-* Right-click the project in Solution Explorer and select **Add** > **Docker Support**.
+* Right-click the project in **Solution Explorer** and select **Add** > **Docker Support**.
+
+## Dockerfile overview
+
+A *Dockerfile*, the recipe for creating a final Docker image, is added to the project root. Refer to [Dockerfile reference](https://docs.docker.com/engine/reference/builder/) for an understanding of the commands within it. This particular *Dockerfile* uses a [multi-stage build](https://docs.docker.com/engine/userguide/eng-image/multistage-build/) with four distinct, named build stages:
+
+::: moniker range=">= aspnetcore-2.1"
+
+[!code-dockerfile[](visual-studio-tools-for-docker/samples/2.1/HelloDockerTools/Dockerfile.original?highlight=1,6,14,17)]
```
Comment (guardrex) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: Note in passing ... I'm glad to see you do this. I've wondered if we should adopt (where needed and with restricted use) code languages that aren't OOB from DocFX.

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker/samples/2.0/HelloDockerTools/Pages/Index.cshtml [hunk 12]
```diff
@@ -1,7 +1,7 @@
 ﻿@page
 @model IndexModel
 @{
-    ViewData["Title"] = "Home page!!!!!";
```
Comment (guardrex) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker/samples/2.0/HelloDockerTools/Pages/Index.cshtml]: Thank you. :+1:

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker/samples/2.1/HelloDockerTools/Startup.cs [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.Configure<CookiePolicyOptions>(options =>
+            {
+                // This lambda determines whether user consent for non-essential cookies is needed for a given request.
+                options.CheckConsentNeeded = context => true;
+                options.MinimumSameSitePolicy = SameSiteMode.None;
+            });
+
+
```
Comment (guardrex) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker/samples/2.1/HelloDockerTools/Startup.cs]: Nit: Remove the extra line space. Is it in the template? I have an open issue with them; so if you know the template is doing this, let me know ... I'll add it to my issue with them.

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md (hunk 7 above)
Comment (scottaddie) [reply aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: I preferred perform too, but Acrolinx didn't.

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md (hunk 8 above)
Comment (scottaddie) [reply aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: I reserve bolding for UI elements. I treat XML elements like code.

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md (hunk 7 above)
Comment (guardrex) [reply aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: *R2-D2 you know better than to trust a strange computer.* -- C-3PO (Star Wars Episode V: The Empire Strikes Back)

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md (hunk 10 above)
Comment (scottaddie) [reply aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: I'll narrow them a bit, but there's only so much that can be done with console output.

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md (hunk 11 above)
Comment (scottaddie) [reply aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: I saw this is one of @JRAlexander's Docker docs and decided to carry it forward.

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md [hunk 14]
```diff
@@ -4,40 +4,42 @@ author: spboyer
...
 
 ## Prerequisites
 
-* [Visual Studio 2017](https://www.visualstudio.com/) with the **.NET Core cross-platform development** workload
 * [Docker for Windows](https://docs.docker.com/docker-for-windows/install/)
+* [Visual Studio 2017](https://www.visualstudio.com/) with the **.NET Core cross-platform development** workload
 
 ## Installation and setup
 
 For Docker installation, review the information at [Docker for Windows: What to know before you install](https://docs.docker.com/docker-for-windows/install/#what-to-know-before-you-install) and install [Docker For Windows](https://docs.docker.com/docker-for-windows/install/).
```
Comment (lisaguthrie) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: I first read this as "review the information at [link] and [link]." As in, review two different pieces of information and then we'll get to the important stuff. So I skipped down to the next paragraph, not realizing that the "install Docker for Windows" piece was actually quite important. I'd suggest: "For Docker installation, first review the information at Docker for Windows: What to know before you install. Then install Docker For Windows." (with appropriate links, etc.)

Comment (lisaguthrie) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: I first read this as "review the information at [link] and [link]." As in, review two different pieces of information and then we'll get to the important stuff. So I skipped down to the next paragraph, not realizing that the "install Docker for Windows" piece was actually quite important. I'd suggest: "For Docker installation, first review the information at Docker for Windows: What to know before you install. Then install Docker For Windows." (with appropriate links, etc.)

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker/samples/2.0/HelloDockerTools/Pages/Index.cshtml (hunk 12 above)
Comment (scottaddie) [reply aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker/samples/2.0/HelloDockerTools/Pages/Index.cshtml]: We need to have a little fun with the samples, right?

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md [hunk 15]
```diff
@@ -46,53 +48,119 @@ If the target framework is .NET Core, the **OS** drop-down allows for the select
 The Visual Studio Tools for Docker don't support adding Docker to an existing ASP.NET Core project targeting .NET Framework. For ASP.NET Core projects targeting .NET Core, there are two options for adding Docker support via the tooling. Open the project in Visual Studio, and choose one of the following options:
```
Comment (lisaguthrie) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: I suggest doing the positive (what IS supported) first :) So, "For ASP.NET Core projects..." and then bullet list that follows. Then the piece about ASP.NET Core targeting .NET Framework not being supported, as its own paragraph after the bullet list.

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md [hunk 16]
```diff
@@ -46,53 +48,119 @@ If the target framework is .NET Core, the **OS** drop-down allows for the select
...
+## Add container orchestrator support to an app
 
-A *Dockerfile*, the recipe for creating a final Docker image, is added to the project root. Refer to [Dockerfile reference](https://docs.docker.com/engine/reference/builder/) for an understanding of the commands within it. This particular *Dockerfile* uses a [multi-stage build](https://docs.docker.com/engine/userguide/eng-image/multistage-build/) containing four distinct, named build stages:
+Visual Studio 2017 versions 15.7 or earlier support [Docker Compose](https://docs.docker.com/compose/overview/) as the sole container orchestration solution. The Docker Compose artifacts are added via **Add** > **Docker Support**.
 
-[!code-dockerfile[](visual-studio-tools-for-docker/samples/HelloDockerTools/HelloDockerTools/Dockerfile?highlight=1,5,14,17)]
+Visual Studio 2017 versions 15.8 or later add an orchestration solution only when instructed. Right-click the project in **Solution Explorer** and select **Add** > **Container Orchestrator Support**. Two different choices are offered: [Docker Compose](#docker-compose) and [Service Fabric](#service-fabric).
```
Comment (lisaguthrie) [inline aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: Those two options are the ones available in box, but extensions can add more options. In particular, the Visual Studio Kubernetes Tools add a Kubernetes option. And I believe we're working on (or already have) an extension for Service Fabric Mesh, which will be exposed through that same dialog box.

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker/samples/2.1/HelloDockerTools/Startup.cs (hunk 13 above)
Comment (scottaddie) [reply aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker/samples/2.1/HelloDockerTools/Startup.cs]: The 2.1 Razor Pages template adds this line space. Here's the complete `Startup.ConfigureServices` method it produces:

```csharp
        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            services.Configure<CookiePolicyOptions>(options =>
            {
                // This lambda determines whether user consent for non-essential cookies is needed for a given request.
                options.CheckConsentNeeded = context => true;
                options.MinimumSameSitePolicy = SameSiteMode.None;
            });


            services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_1);
        }
```

Comment (guardrex) [reply aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker/samples/2.1/HelloDockerTools/Startup.cs]: kk ... I let them know about the extra line space here.

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md (hunk 15 above)
Comment (scottaddie) [reply aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker.md]: You're absolutely right. It's best to lead with the positive stuff.

File: aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker/samples/2.0/HelloDockerTools/Pages/Index.cshtml (hunk 12 above)
Comment (guardrex) [reply aspnetcore/host-and-deploy/docker/visual-studio-tools-for-docker/samples/2.0/HelloDockerTools/Pages/Index.cshtml]: 🤦‍♂️ I need to *eat my remark now*. I just discovered in a Blazor sample that I told the reader to add exclamation points to a heading in a sample app ... Do what I *say*, not what I *do*. That's the advice I follow ..... I only do what I *say* .... I never *do* what I *do*. :smile:

39509668-e8949a92-4d9b-11e8-91e9-b6a494695d92

**Pull request merged**